test/cpp_api_parity/module_impl_check.py

```diff
@@ -0,0 +1,293 @@
+# The purpose of this test is to check that we have implementation parity between
+# a Python `torch.nn` module and its corresponding C++ `torch::nn` module. Concretely,
+# this test does the following:
+#
+# 1. Get a test params dict from common_nn.py, run forward and backward on the
+# Python module created using the test params.
+#
+# 2. Serialize the Python module's parameters / buffers and its forward input
+# arguments, deserialize them in C++ and load them into the C++ module.
+#
+# 3. Run the same forward and backward passes on the C++ module, and serialize
+# the C++ module's forward output and backward gradients.
+#
+# 4. Compare Python/C++ module's forward output and backward gradients. If they
+# are the same, then we have implementation parity between Python/C++ module.
+
+import tempfile
+import shutil
+from string import Template
+import types
+
+import torch
+from cpp_api_parity.utils import TorchNNModuleTestParams, TORCH_NN_COMMON_TEST_HARNESS, \
+    compile_cpp_code_inline, set_python_tensors_requires_grad, move_python_tensors_to_device, \
+    add_test, compute_cpp_args_construction_stmts_and_forward_arg_symbols, serialize_arg_dict_as_script_module, \
+    compute_arg_dict, decorate_test_fn, compute_temp_file_path, generate_error_msg
+from cpp_api_parity import torch_nn_modules
+
+# Expected substitutions:
+#
+# ${module_variant_name}
+# ${module_qualified_name}
+# ${cpp_tmp_folder}
+# ${cpp_args_construction_stmts}
+# ${cpp_constructor_args}
+# ${device}
```

Comment:
Do we have to use device as a template?

Reply:
I think we have two options: 1. use device as a template, 2. pass in device as a function argument. The first option is probably better because we already encode device information (i.e. `cpu` vs `cuda`)  in `module_variant_name`, and so it is probably not a good idea to allow passing in a different device as function argument.

Reply:
Yup, that sounds reasonable, thanks for the explanation